### DISCUSSION TITLE
Equal block proposing opportunities for all validators in the active set

### DISCUSSION THREAD

**Discussion by kwunyeung** [desmos-labs/desmos#365] [Features ideas]
One of the features in our proposed tokenomics is to let all validators have the same chance of proposing block so that the proposer bonus will and rewards will be distributed in in the same chance. The idea is to break the relationship between number of delegated tokens and proposing priority. 

To achieve this, there are two approaches.

1. We set voting power of all validators in the active set to be equal. This is how IOV works. And override how the [delegation](https://github.com/cosmos/cosmos-sdk/blob/master/x/staking/spec/02_state_transitions.md#delegations) in the Cosmos SDK which it will not update the voting power based on the delegated tokens. In this approach, each proposer will have equal weight of chance in proposing blocks based on the current proposer priority mechanism.

2. A long term solution is to apply the random selection of block proposer by using [Verifiable Random Function](http://cryptowiki.net/index.php?title=Verifiable_Random_Functions). This is current currently [Algorand is using for selecting consensus committee](https://medium.com/algorand/algorand-releases-first-open-source-code-of-verifiable-random-function-93c2960abd61). The approach has also been discussed in [Tendermint](https://github.com/tendermint/tendermint/issues/763).

**Comment by kwunyeung** [comment]
We may try the following approach. On every block, take the block hash as seed and generate a deterministic random number by picking the first result from the sequence.

**Comment by kwunyeung** [comment]
I have read the [article form CodeChain](https://medium.com/codechain/tenderand-randomized-leader-election-in-tendermint-a3663d863479) but this is not what we would like to experiment. Although the proposer selection is random but they are still in proportion to the stake volume. What we are go to experiment is to see if an even block producing mechanism with high block proposer fee bonus will help the fee distribution more even to reduce the problem of rich get richer.

**Comment by kwunyeung** [comment]
@RiccardoM @bragaz The block header of Tendermint does have `Time`, and `AppHash` in it.

https://github.com/desmos-labs/tendermint/blob/master/types/block.go#L328

Manipulating everything to come up with a hash for higher proposer priority is just too difficult unless the bad validator has very high power which can control the consensus mean time.

**Comment by rmontagnin** [comment]
@kwunyeung So do you think the way you've thought about generating the next proposer is sufficiently safe?

**Comment by leobragaz** [comment]
> unless the bad validator has very high power which can control the consensus mean time

With a random selected proposer this scenario shouldn't be a problem right?
(and, even if it would be, couldn't we mitigate it setting the same voting power for all validators?)

**Comment by kwunyeung** [comment]
@RiccardoM practically safe as I don't expect too much variance on voting power on `Desmos` but theoretically in computer science perspective it is not safe from hash grinding.

We should studying the BLS random beacon approach to incorporate VRF which P2P has been done this this version of `Tendermint`. The idea is to get the threshold signature from all validators to generate a new seed for the randomization.

https://economy.p2p.org/p2p-validator-helps-to-carry-out-random-beacon-for-Cosmos/

@bragaz I would not say setting same voting power but "The Foundation" which would hold quite some amount of tokens and we can help balancing that situation on mainnet. I still want to have a way that the validators to do community work and market to attract external delegators.

Although the existing approach is not the best solution, I think we can use this to start testing the dynamics of the fee distribution to validators when we can randomize the block proposer. This will help us study how the fee distribution should be set in the `x/distribution`. Many existing networks are not sustainable as the income of the validators rely on **`INFLATION`** which most of the tokens are distributed to the top validators. Our hypothesis is that if large portion of **`TRANSACTION FEE`** can be shared among validators and if we have enough amount of transaction, the validators can survive and the network is sustainable. We can study this in the meantime see how a better solution can be adapted.

**Comment by kwunyeung** [comment]
`Octracon` from LINE as done the random proposer selection using VRF https://github.com/line/ostracon/pull/194. It has now based on `Tendermint Core v0.34.8`. Although randomization is still based on voting power, it's more secured than selecting proposer in round-robin and it allows using BLS which solve some validators signature verifications bottleneck.
